osgar/drivers/eduro.py

```diff
@@ -0,0 +1,228 @@
+from osgar.bus import BusShutdownException
+from .canserial import CAN_packet
+
+CAN_ID_SYNC = 0x80
+CAN_ID_ENCODERS_LEFT = 0x181
+CAN_ID_ENCODERS_RIGHT = 0x182
+CAM_SYSTEM_STATUS = 0x8A  # including emergency STOP button
+CAN_ID_BUTTONS = 0x28A
+
+UPDATE_TIME_FREQUENCY = 20.0  # Hz
+WHEEL_DIAMETER_LEFT = 427.0 / 445.0 * 0.26/4.0 + 0.00015
+WHEEL_DIAMETER_RIGHT = 427.0 / 445.0 * 0.26/4.0 - 0.00015
+
+
+def sint32_diff(a, b):
+    return ctypes.c_int32(a - b).value
+
+
+class Eduro(Thread):
+    def __init__(self, config, bus):
+        Thread.__init__(self)
+        self.setDaemon(True)
+
+        self.bus = bus
+
+        self.desired_speed = 0.0  # m/s
+        self.desired_angular_speed = 0.0
+        self.maxAcc = config.get('max_acc', 0.5)
+        self._rampLastLeft, self._rampLastRight = 0.0, 0.0
+        self.WHEEL_DISTANCE = 0.315        
+
+        self.prev_enc_raw = {}
+        self.dist_left_raw = 0
+        self.dist_right_raw = 0
+        self.dist_pose = 0, 0
+
+        self.emergency_stop = None  # uknown state
+        self.pose = [0.0, 0.0, 0.0]  # x, y in meters, heading in radians (not corrected to 2PI)
+        self.buttons = None  # unknown
+
+    def update_encoders(self, msg_id, data):
+        assert len(data) == 4, data
+        arr = struct.unpack('<i', data)
+        if msg_id in self.prev_enc_raw:
+            diff = sint32_diff(arr[0], self.prev_enc_raw[msg_id])
+
+            if msg_id == CAN_ID_ENCODERS_LEFT:
+                self.dist_left_raw += diff
+            elif msg_id == CAN_ID_ENCODERS_RIGHT:
+                self.dist_right_raw += diff
+        self.prev_enc_raw[msg_id] = arr[0]
+
+    def update_emergency_stop(self, msg_id, data):
+        assert len(data) == 8, len(data)
+        self.emergency_stop = (data[:2] == bytes([0,0x10])) and (data [3:] == bytes([0,0,0,0,0]))
+
+    def update_pose(self):
+        x, y, heading = self.pose
+
+        diff_left = self.dist_left_raw - self.dist_pose[0]
+        diff_right = self.dist_right_raw  - self.dist_pose[1]
+        self.dist_pose = self.dist_left_raw, self.dist_right_raw
+
+        scaleR = WHEEL_DIAMETER_RIGHT * math.pi/0x10000
+        scaleL = WHEEL_DIAMETER_LEFT * math.pi/0x10000
+
+        metricL = scaleL * diff_left
+        metricR = scaleR * diff_right
+
+        dist = (metricL + metricR)/2.0
+        angle = (metricR - metricL)/self.WHEEL_DISTANCE
+
+        # advance robot by given distance and angle
+        if abs(angle) < 0.0000001:  # EPS
+            # Straight movement - a special case
+            x += dist * math.cos(heading)
+            y += dist * math.sin(heading)
+            #Not needed: heading += angle
+        else:
+            # Arc
+            r = dist / angle
+            x += -r * math.sin(heading) + r * math.sin(heading + angle)
+            y += +r * math.cos(heading) - r * math.cos(heading + angle)
+            heading += angle # not normalized
+        self.pose = [x, y, heading]
```

Comment:
If `self.pose` is never updated in place, can it be a tuple?

Reply:
well, I used to use `tuple` but with `msgpack` I switched to `list` ... but then I changed again to integer values ... so I can change it back to tuple ... OK